core/src/main/scala/org/apache/spark/scheduler/JobWaiter.scala

```diff
@@ -49,29 +51,14 @@ private[spark] class JobWaiter[T](
     dagScheduler.cancelJob(jobId)
   }
 
-  override def taskSucceeded(index: Int, result: Any): Unit = synchronized {
-    if (_jobFinished) {
-      throw new UnsupportedOperationException("taskSucceeded() called on a finished JobWaiter")
-    }
+  override def taskSucceeded(index: Int, result: Any): Unit = {
```

Comment:
I wonder... do we need to also put some sort of synchronization around `resultFunc` in `SimpleFutureAction` to ensure that the thread that reads the result can see all the writes?